File: apollo-api/src/test/java/com/apollographql/apollo/api/InputTest.java
package com.apollographql.apollo.api;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

public final class InputTest {


    @Test
    public void testInputEqualsOnNotNullValue() {
        String value = "Hello world!";
        Input<String> stringInput = Input.fromNullable(value);
        Input<String> anotherStringInput = Input.fromNullable(value);

        assertEquals(stringInput, anotherStringInput);
    }

    @Test
    public void testInputNotEqualsOnDifferentValues() {
        String value = "Hello world!";
        String value2 = "Bye world!";
        Input<String> stringInput = Input.fromNullable(value);
        Input<String> anotherStringInput = Input.fromNullable(value2);

        assertNotEquals(stringInput, anotherStringInput);
    }

    @Test
    public void testInputEqualsOnNullValue() {
        Input<String> stringInput = Input.fromNullable(null);
        Input<String> anotherStringInput = Input.fromNullable(null);

        assertEquals(stringInput, anotherStringInput);
    }

    @Test
    public void testInputEqualsOnNotNullObjects() {
        TestObject object = new TestObject("Hello world!");
        Input<TestObject> aInput = Input.fromNullable(object);
        Input<TestObject> anotherInput = Input.fromNullable(object);

        assertEquals(aInput, anotherInput);
    }

    @Test
    public void testInputEqualsOnEqualObjectsWithDifferentReferences() {

        TestObject object1 = new TestObject("Hello world!");
        TestObject object2 = new TestObject("Hello world!");
        Input<TestObject> input1 = Input.fromNullable(object1);
        Input<TestObject> input2 = Input.fromNullable(object2);
        assertEquals(input1, input2);
    }

    @Test
    public void testInputNotEqualsOnDifferentObjects() {
        TestObject object = new TestObject("Hello world!");
        TestObject anotherObject = new TestObject("Bye world!");
        Input<TestObject> aInput = Input.fromNullable(object);
        Input<TestObject> anotherInput = Input.fromNullable(anotherObject);

        assertNotEquals(aInput, anotherInput);
    }

    @Test
    public void testInputEqualsOnObjectsWithNullValue() {
        TestObject object = new TestObject(null);
        Input<TestObject> aInput = Input.fromNullable(object);
        Input<TestObject> anotherInput = Input.fromNullable(object);

        assertEquals(aInput, anotherInput);
    }

    @Test
    public void testInputNotEqualsWhenAnObjectIsNull() {
        TestObject object = new TestObject(null);
        Input<TestObject> aInput = Input.fromNullable(object);
        Input<TestObject> anotherInput = Input.fromNullable(null);

        assertNotEquals(aInput, anotherInput);
    }

    //==================================================================
    //==================================================================
    class TestObject {

        private String value;

        TestObject(String value) {
            this.value = value;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof TestObject)) {
                return false;
            }

            TestObject input = (TestObject) o;
            return value != null && value.equals(input.value) || value == null && input.value == null;
        }

      @Override
      public int hashCode() {
        return super.hashCode();
      }
    }
}

File: apollo-api/src/commonTest/kotlin/com/apollographql/apollo3/api/InputTest.kt
package com.apollographql.apollo3.api

import com.apollographql.apollo3.api.Input.Companion.fromNullable
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals

class InputTest {
  @Test
  fun testInputEqualsOnNotNullValue() {
    val value = "Hello world!"
    val stringInput = fromNullable(value)
    val anotherStringInput = fromNullable(value)
    assertEquals(stringInput, anotherStringInput)
  }

  @Test
  fun testInputNotEqualsOnDifferentValues() {
    val value = "Hello world!"
    val value2 = "Bye world!"
    val stringInput = fromNullable(value)
    val anotherStringInput = fromNullable(value2)
    assertNotEquals(stringInput, anotherStringInput)
  }

  @Test
  fun testInputEqualsOnNullValue() {
    val stringInput = fromNullable<String>(null)
    val anotherStringInput = fromNullable<String>(null)
    assertEquals(stringInput, anotherStringInput)
  }

  @Test
  fun testInputEqualsOnNotNullObjects() {
    val `object` = TestObject("Hello world!")
    val aInput = fromNullable(`object`)
    val anotherInput = fromNullable(`object`)
    assertEquals(aInput, anotherInput)
  }

  @Test
  fun testInputEqualsOnEqualObjectsWithDifferentReferences() {
    val object1 = TestObject("Hello world!")
    val object2 = TestObject("Hello world!")
    val input1 = fromNullable(object1)
    val input2 = fromNullable(object2)
    assertEquals(input1, input2)
  }

  @Test
  fun testInputNotEqualsOnDifferentObjects() {
    val `object` = TestObject("Hello world!")
    val anotherObject = TestObject("Bye world!")
    val aInput = fromNullable(`object`)
    val anotherInput = fromNullable(anotherObject)
    assertNotEquals(aInput, anotherInput)
  }

  @Test
  fun testInputEqualsOnObjectsWithNullValue() {
    val `object` = TestObject(null)
    val aInput = fromNullable(`object`)
    val anotherInput = fromNullable(`object`)
    assertEquals(aInput, anotherInput)
  }

  @Test
  fun testInputNotEqualsWhenAnObjectIsNull() {
    val `object` = TestObject(null)
    val aInput = fromNullable(`object`)
    val anotherInput: Input<TestObject> = fromNullable(null)
    assertNotEquals(aInput, anotherInput)
  }

  //==================================================================
  //==================================================================
  internal data class TestObject(private val value: String?)
}
